Provide a way to set timeouts for GCM, APNS and MPNS + sane defaults

elhu opened the pull request:
Hello!

This PR addresses problem we've faced a bunch of time in production, where we've had Sidekiq workers that get stuck sending notifications.
Since no timeout is provided, the job stays stuck until the next worker restart, at which point the notification has long since become useless.

Technically the change wasn't necessary for APNS, but this way the signature of each `XYZPusher` initializer is consistent and timeout is always configurable.

calonso commented on examples/apns.rb:
```diff
@@ -13,6 +13,9 @@
 
 pusher = RubyPushNotifications::APNS::APNSPusher.new(File.read('/path/to/your/apps/certificate.pem'), true, password) # enter the password if present
+# Connect timeout defaults to 30s
+# pusher = RubyPushNotifications::APNS::APNSPusher.new(File.read('/path/to/your/apps/certificate.pem'), true, password, connect_timeout: { 20 })
```
`connect_timeout: { 20 }` doesn't make much sense, did you meant `{ connect_timeout: 20 }`? :)

calonso commented on lib/ruby-push-notifications/apns/apns_connection.rb:
```diff
@@ -26,14 +26,16 @@ class APNSConnection
       # @param cert [String]. Contents of the PEM encoded certificate
       # @param sandbox [Boolean]. Whether to use the sandbox environment or not.
+      # @param optional pass [String]. Passphrase for the certificate.
+      # @param optional options [Hash]. Options for the TCP socket
```
Options are not only for the TCP socket, it could be used for many more things later on. Also I'd add a list of currently used fields.

calonso commented on lib/ruby-push-notifications/apns/apns_pusher.rb:
```diff
@@ -15,10 +15,12 @@ class APNSPusher
       # @param certificate [String]. The PEM encoded APNS certificate.
       # @param sandbox [Boolean]. Whether the certificate is an APNS sandbox or not.
-      def initialize(certificate, sandbox, password = nil)
+      # @param optional options [Hash]. Options for the TCP socket.
```
Same as in `APNSConnection`

calonso commented on lib/ruby-push-notifications/gcm/gcm_connection.rb:
```diff
@@ -27,8 +27,9 @@ class GCMConnection
       # @param notification [String]. The text to POST
       # @param key [String]. The GCM sender id to use
       #    (https://developer.android.com/google/gcm/gcm.html#senderid)
+      # @param optional options [Hash]. Options for the HTTP connection.
```
Same as before. It would be helpful to add a list of the valid values and defaults

calonso commented on lib/ruby-push-notifications/mpns/mpns_pusher.rb:
```diff
@@ -21,7 +23,7 @@ def initialize(certificate = nil)
       def push(notifications)
         notifications.each do |notif|
-          notif.results = MPNSConnection.post notif, @certificate
+          notif.results = MPNSConnection.post notif, @certificate, options
```
`options` or `@options`?

calonso commented:
Also we'll need to bump the gem minor version I guess.

elhu commented:
I addressed most of your comments (took the opportunity to fix some YARD formatting).
The main problem now is the reason the CI fails.

The signature of `Socket.tcp` changed in Ruby 2.0 (2 years and 10 months ago).
It did not used to accept options (http://ruby-doc.org/stdlib-1.9.3/libdoc/socket/rdoc/Socket.html#method-c-tcp) but now it does (http://ruby-doc.org/stdlib-2.0.0/libdoc/socket/rdoc/Socket.html#method-c-tcp)

I could try to find a way to circumvent the problem, but any fix might be pretty ugly. 
Is supporting ruby-1.9.3 a hard requirement for you or can I drop it from the travis build matrix and add the version requirement to the gemspec?

calonso commented:
Ok, I see... Well, according to [some stats](http://rails-hosting.com/2016/index.html) Ruby 1.9.3 is still present, but I'm not sure if the percentage of users is relevant, and also I'd be interested in knowing who is using this gem and which Ruby versions they're using...

Let's try to keep it compatible, as there's only one affected point (`APNSConnection`) but if you find it gets hairy then I don't really have strong requirements on keeping 1.9.3.

Thanks a lot for your good work here.

elhu commented:
The "best" way to do this pre-2.0.0 is apparently quite convoluted (see the accepted answer on http://stackoverflow.com/questions/21014272/how-ruby-tcpsocket-timeout-is-defined).

The way I see it, we can either:
1. drop my changes to APNS and let the default `connect_timeout` of `TCPSocket.new` do its job
2. drop support for ruby 1.9.3.

If we go with #2, we could keep the new `options` parameter to `RubyPushNotifications::APNS::APNSPusher.initialize` to keep the method signature consistent with the rest.

For the record, I just installed 1.9.3 to make sure that the default timeout works (it does), and here's the install message:
```
WARNING: ruby-1.9.3-p551 is past its end of life and is now unsupported.
It no longer receives bug fixes or critical security updates.
```

calonso commented:
Ok, let's drop it then as this won't stop people on 1.9.3 from using it. It's just that they won't be fully featured/supported, but they're already on that situation so I think is a fair decision.

elhu commented:
OK, I've dropped the support for ruby 1.9.3. I'll let you be the judge of whether or not this warrants a major version bump!

calonso commented:
Ok, cool. So according to the semantic versioning principles, major version 0 (in which we are right now) means that we're on development stage and thus anything from the public API may be changed at any time so, according to it, a minor version bump should suffice, on the other hand, I wonder if this is stable enough to consider it production ready by releasing 1.0.0 and for that decision your opinion really counts. How do you feel using this gem? How long have you been using it? Suggestions? Concerns?

elhu commented:
Full disclosure, we're only using this for GCM. I've applied the same timeout logic to the other two Push notification providers to keep the PR consistent and also maybe help a few more users.

We've been using it in production since May 2015. No problem so far, everything is stable and plenty fast enough for our use case, so I don't see any problem with releasing 1.0.0.

calonso commented:
Ok coo, let's go 1.0.0 then.
Cheers!

elhu commented:
Cool, I've bumped the version in the gemspec!
Do you need anything more before you can merge?

elhu commented:
Thank you for merging this! Do you know when you'll be releasing the gem?

calonso commented:
Yeah, it is done right now [https://rubygems.org/gems/ruby-push-notifications/versions/1.0.0](https://rubygems.org/gems/ruby-push-notifications/versions/1.0.0).

Thank you very much for your awesome work. BTW, I'm adding you to the collaborators list. 
Welcome to the team!!
